### PR TITLE
[Resource] Network Security Group Rule 

### PR DESCRIPTION
**Pull request by ismirlia**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

### CODE REVIEW

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go [hunk 1]
```diff
@@ -0,0 +1,611 @@
...
+	networkSecurityGroupRuleRemote.ID = remoteMap[Arg_ID].(string)
+	networkSecurityGroupRuleRemote.Type = remoteMap[Arg_Type].(string)
+	return &networkSecurityGroupRuleRemote
+}
+
+func networkSecurityGroupRuleMapToProtocol(protocolMap map[string]interface{}) *models.NetworkSecurityGroupRuleProtocol {
+	networkSecurityGroupRuleProtocol := models.NetworkSecurityGroupRuleProtocol{}
+	networkSecurityGroupRuleProtocol.Type = protocolMap[Arg_Type].(string)
+
+	if networkSecurityGroupRuleProtocol.Type == "icmp" {
```
Comment (michaelkad) [inline ibm/service/power/resource_ibm_pi_network_security_group_rule.go:596]: Could you add a const for these ? `"icmp", "tcp"`

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go [hunk 2]
```diff
@@ -0,0 +1,611 @@
...
+func networkSecurityGroupRuleMapToPort(portMap map[string]interface{}) *models.NetworkSecurityGroupRulePort {
+	networkSecurityGroupRulePort := models.NetworkSecurityGroupRulePort{}
+	networkSecurityGroupRulePort.Maximum = portMap[Attr_Maximum].(float64)
+	networkSecurityGroupRulePort.Minimum = portMap[Attr_Minimum].(float64)
+	return &networkSecurityGroupRulePort
+}
+
+func networkSecurityGroupRuleMapToRemote(remoteMap map[string]interface{}) *models.NetworkSecurityGroupRuleRemote {
+	networkSecurityGroupRuleRemote := models.NetworkSecurityGroupRuleRemote{}
+	networkSecurityGroupRuleRemote.ID = remoteMap[Arg_ID].(string)
```
Comment (michaelkad) [inline ibm/service/power/resource_ibm_pi_network_security_group_rule.go:587]: Could you add a check for `remoteMap[Arg_ID]`? since this can be empty

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go [hunk 3]
```diff
@@ -0,0 +1,611 @@
...
+	if remote.Type != "" {
+		remoteMap[Attr_Type] = remote.Type
+	}
+	return remoteMap
+}
+
+func networkSecurityGroupRuleMapToPort(portMap map[string]interface{}) *models.NetworkSecurityGroupRulePort {
+	networkSecurityGroupRulePort := models.NetworkSecurityGroupRulePort{}
+	networkSecurityGroupRulePort.Maximum = portMap[Attr_Maximum].(float64)
+	networkSecurityGroupRulePort.Minimum = portMap[Attr_Minimum].(float64)
```
Comment (michaelkad) [inline ibm/service/power/resource_ibm_pi_network_security_group_rule.go:581]: Could you add a check here for those values?

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go [hunk 4]
```diff
@@ -0,0 +1,611 @@
...
+					return networkSecurityGroup, State_Pending, nil
+				}
+			}
+			if !isMember {
+				return networkSecurityGroup, State_Removed, nil
+			}
+		}
+		return networkSecurityGroup, State_Pending, nil
+	}
+}
```
Comment (michaelkad) [inline ibm/service/power/resource_ibm_pi_network_security_group_rule.go:502]: Why are you checking for member if this is after removing a rule?

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go [hunk 5]
```diff
@@ -0,0 +1,611 @@
...
+			for _, mbr := range networkSecurityGroup.Members {
+				if *mbr.ID == memberID {
+					return networkSecurityGroup, State_Available, nil
+				}
+
+			}
+		}
+		return networkSecurityGroup, State_Pending, nil
+	}
+}
```
Comment (michaelkad) [inline ibm/service/power/resource_ibm_pi_network_security_group_rule.go:464]: Why are you checking for member if this is after adding a rule?

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go [hunk 6]
```diff
@@ -0,0 +1,611 @@
...
+	nsgClient := instance.NewIBMIPINetworkSecurityGroupClient(ctx, sess, cloudInstanceID)
+	nsgID := d.Get(Arg_NetworkSecurityGroupID).(string)
+
+	if v, ok := d.GetOk(Arg_NetworkSecurityGroupRuleID); ok {
+		ruleID := v.(string)
+		_, err := nsgClient.DeleteRule(nsgID, ruleID)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		_, err = isWaitForIBMPINetworkSecurityGroupRuleRemove(ctx, nsgClient, nsgID, ruleID, d.Timeout(schema.TimeoutDelete))
```
Comment (michaelkad) [inline ibm/service/power/resource_ibm_pi_network_security_group_rule.go:346]: Left a comment for this `isWaitForIBMPINetworkSecurityGroupRuleRemove` below

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go [hunk 7]
```diff
@@ -0,0 +1,611 @@
...
+		networkSecurityGroupAddRule.DestinationPorts = networkSecurityGroupRuleMapToPort(destinationPort)
+
+		sourcePort := d.Get(Arg_SourcePorts).(map[string]interface{})
+		networkSecurityGroupAddRule.SourcePorts = networkSecurityGroupRuleMapToPort(sourcePort)
+
+		networkSecurityGroup, err := nsgClient.AddRule(nsgID, &networkSecurityGroupAddRule)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		_, err = isWaitForIBMPINetworkSecurityGroupRuleAdd(ctx, nsgClient, nsgID, *networkSecurityGroup.ID, d.Timeout(schema.TimeoutCreate))
```
Comment (michaelkad) [inline ibm/service/power/resource_ibm_pi_network_security_group_rule.go:379]: Left a comment for this `isWaitForIBMPINetworkSecurityGroupRuleAdd` below

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go [hunk 8]
```diff
@@ -0,0 +1,611 @@
...
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		d.SetId(fmt.Sprintf("%s/%s/%s", cloudInstanceID, nsgID, *networkSecurityGroup.ID))
```
Comment (michaelkad) [inline ibm/service/power/resource_ibm_pi_network_security_group_rule.go:383]: `nsgID == *networkSecurityGroup.ID` We want the rule id that was just created but we can't since the api does not return it :sweat:

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go [hunk 9]
```diff
@@ -0,0 +1,611 @@
...
+						Arg_Minimum: {
+							Computed:    true,
+							Description: "The start of the port range, if applicable. If values are not present then all ports are in the range.",
+							Type:        schema.TypeFloat,
+						},
+					},
+				},
+				Optional: true,
+				Type:     schema.TypeList,
+			},
```
Comment (michaelkad) [inline ibm/service/power/resource_ibm_pi_network_security_group_rule.go:67]: I think you can keep the top level as `Arg_...` and the inner one as `Attr_...`

```suggestion
			Arg_DestinationPorts: {
				Computed:      true,
				ConflictsWith: []string{Arg_NetworkSecurityGroupRuleID},
				Description:   "Destination port ranges.",
				Elem: &schema.Resource{
					Schema: map[string]*schema.Schema{
						Att_Maximum: {
							Computed:    true,
							Description: "The end of the port range, if applicable. If values are not present then all ports are in the range.",
							Type:        schema.TypeFloat,
						},
						Att_Minimum: {
							Computed:    true,
							Description: "The start of the port range, if applicable. If values are not present then all ports are in the range.",
							Type:        schema.TypeFloat,
						},
					},
				},
				Optional: true,
				Type:     schema.TypeList,
			},
```

**Review by michaelkad — CHANGES_REQUESTED** [review]
The title is correct, should be `Rule` instead of `Member`.

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go (hunk 9 above)
Comment (ismirlia) [reply ibm/service/power/resource_ibm_pi_network_security_group_rule.go:67]: But doesn't the user, provide those values, shouldn't there be a Arg and Attr version?

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go (hunk 1 above)
Comment (ismirlia) [reply ibm/service/power/resource_ibm_pi_network_security_group_rule.go:596]: Updated

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go (hunk 4 above)
Comment (ismirlia) [reply ibm/service/power/resource_ibm_pi_network_security_group_rule.go:502]: I copy pasted and forgot to replace them.

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go (hunk 5 above)
Comment (ismirlia) [reply ibm/service/power/resource_ibm_pi_network_security_group_rule.go:464]: I copy pasted and forgot to replace them, should be checking for the new rule id.

File: website/docs/r/pi_network_security_group_rule.html.markdown [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+- `pi_remote` - (Optional, List) List of remote. Required if `pi_network_security_group_rule_id` is not provided.
+
+    Nested schema for `pi_remote`:
+    - `pi_id` - (Optional, String) The id of the remote network address group or network security group the rules apply to. Not required for default-network-address-group.
+    - `pi_type` - (Optional, String) The type of remote group the rules apply to. Supported values are: `network-security-group`, `network-address-group`, `default-network-address-group`.
+- `pi_source_port` - (Optional, List) List of source port
+
+    Nested schema for `pi_source_port`:
+    - `pi_maximum` - (Optional, Int) The end of the port range, if applicable, If values are not present then all ports are in the range.
+    - `pi_minimum` - (Optional, Int) The start of the port range, if applicable. If values are not present then all ports are in the range.
```
Comment (Alexander-Kita) [inline website/docs/r/pi_network_security_group_rule.html.markdown:66]:
```suggestion
    - `pi_maximum` - (Optional, Int) The end of the port range, if applicable. If the value is not present then the default value of 65535 will be the maximum port number.
    - `pi_minimum` - (Optional, Int) The start of the port range, if applicable. If the value is not present then the default value of 1 will be the minimum port number.
```

File: website/docs/r/pi_network_security_group_rule.html.markdown [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+Review the argument references that you can specify for your resource.
+
+- `pi_action` - (Optional, String) The action to take if the rule matches network traffic. Supported values are: `allow`, `deny`. Required if `pi_network_security_group_rule_id` is not provided.
+- `pi_cloud_instance_id` - (Required, String) The GUID of the service instance associated with an account.
+- `pi_destination_port` - (Optional, List) The list of destination port.
+
+    Nested schema for `pi_destination_port`:
+      - `pi_maximum` - (Optional, Int) The end of the port range, if applicable, If values are not present then all ports are in the range.
+      - `pi_minimum` - (Optional, Int) The start of the port range, if applicable. If values are not present then all ports are in the range.
```
Comment (Alexander-Kita) [inline website/docs/r/pi_network_security_group_rule.html.markdown:48]:
```suggestion
      - `pi_maximum` - (Optional, Int) The end of the port range, if applicable. If the value is not present then the default value of 65535 will be the maximum port number.
      - `pi_minimum` - (Optional, Int) The start of the port range, if applicable. If the value is not present then the default value of 1 will be the minimum port number.
```

File: website/docs/r/pi_network_security_group_rule.html.markdown [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+- `network_security_group_id` -(String) The unique identifier of the network security group.
+- `rules` - (List) The list of rules in the network security group.
+
+    Nested schema for `rules`:
+  - `action` - (String) The action to take if the rule matches network traffic. Supported values are: `allow`, `deny`.
+  - `destination_port` - (List) The list of destination port.
+
+        Nested schema for `destination_port`:
+          - `maximum` - (Int) The end of the port range, if applicable, If values are not present then all ports are in the range.
+          - `minimum` - (Int) The start of the port range, if applicable. If values are not present then all ports are in the range.
```
Comment (Alexander-Kita) [inline website/docs/r/pi_network_security_group_rule.html.markdown:93]:
```suggestion
          - `maximum` - (Int) The end of the port range, if applicable. If the value is not present then the default value of 65535 will be the maximum port number.
          - `minimum` - (Int)The start of the port range, if applicable. If the value is not present then the default value of 1 will be the minimum port number.
```

File: website/docs/r/pi_network_security_group_rule.html.markdown [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+  - `remote` - (List) List of remote.
+
+        Nested schema for `remote`:
+        - `id` - (String) The id of the remote network address group or network security group the rules apply to. Not required for default-network-address-group.
+        - `type` - (String) The type of remote group the rules apply to. Supported values are: `network-security-group`, `network-address-group`, `default-network-address-group`.
+  - `source_port` - (List) List of source port
+
+        Nested schema for `source_port`:
+        - `maximum` - (Int) The end of the port range, if applicable, If values are not present then all ports are in the range.
+        - `minimum` - (Int) The start of the port range, if applicable. If values are not present then all ports are in the range.
```
Comment (Alexander-Kita) [inline website/docs/r/pi_network_security_group_rule.html.markdown:111]:
```suggestion
        - `maximum` - (Int) The end of the port range, if applicable. If the value is not present then the default value of 65535 will be the maximum port number.
        - `minimum` - (Int) The start of the port range, if applicable. If the value is not present then the default value of 1 will be the minimum port number.
```

File: ibm/service/power/ibm_pi_constants.go [hunk 14]
```diff
@@ -84,6 +98,8 @@ const (
 	Arg_StorageType                         = "pi_storage_type"
 	Arg_SysType                             = "pi_sys_type"
 	Arg_TargetStorageTier                   = "pi_target_storage_tier"
+	Arg_TCPFlags                            = "pi_tcp_flags"
```
Comment (Alexander-Kita) [inline ibm/service/power/ibm_pi_constants.go:101]: This does not appear to be used anywhere

**Comment by ismirlia** [comment]
Unit test results:
```
--- PASS: TestAccIBMPINetworkSecurityGroupRuleBasic (31.59s)
PASS

--- PASS: TestAccIBMPINetworkSecurityGroupRuleTCP (34.92s)
PASS

--- PASS: TestAccIBMPINetworkSecurityGroupRuleRemove (22.49s)
PASS
```

File: website/docs/r/pi_network_security_group_rule.html.markdown [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+        minimum = 1200
+        maximum = 37466
+      }
+      pi_source_ports {
+        minimum = 1000
+        maximum = 19500
+      }
+      pi_protocol {
+        icmp_types = []
+        tcp_flags  = ["syn", "ack"]
```
Comment (Alexander-Kita) [inline website/docs/r/pi_network_security_group_rule.html.markdown:30]: This needs to be updated to the new way to defined TCP flags

File: ibm/service/power/ibm_pi_constants.go [hunk 16]
```diff
@@ -42,26 +46,35 @@ const (
...
 	Arg_PVMInstanceId                       = "pi_instance_id"
+	Arg_Remote                              = "pi_remote"
+	Arg_RemoteID                            = "pi_remote_id"
 	Arg_Remove                              = "pi_remove"
 	Arg_Replicants                          = "pi_replicants"
 	Arg_ReplicationEnabled                  = "pi_replication_enabled"
 	Arg_ReplicationPolicy                   = "pi_replication_policy"
 	Arg_ReplicationScheme                   = "pi_replication_scheme"
 	Arg_ResourceGroupID                     = "pi_resource_group_id"
+	Arg_RuleID                              = "pi_rule_id"
```
Comment (michaelkad) [inline ibm/service/power/ibm_pi_constants.go:76]: This can be removed. Not in use.

File: ibm/service/power/ibm_pi_constants.go [hunk 17]
```diff
@@ -42,26 +46,35 @@ const (
...
+	Arg_RuleName                            = "pi_rule_name"
```
Comment (michaelkad) [inline ibm/service/power/ibm_pi_constants.go:77]: This can be removed. Not in use.

File: ibm/service/power/ibm_pi_constants.go [hunk 18]
```diff
@@ -84,6 +98,8 @@ const (
 	Arg_StorageType                         = "pi_storage_type"
 	Arg_SysType                             = "pi_sys_type"
 	Arg_TargetStorageTier                   = "pi_target_storage_tier"
+	Arg_TCPFlags                            = "pi_tcp_flags"
+	Arg_Type                                = "pi_type"
```
Comment (michaelkad) [inline ibm/service/power/ibm_pi_constants.go:101]: This can be removed. Not in use.

File: ibm/service/power/ibm_pi_constants.go [hunk 19]
```diff
@@ -260,6 +283,10 @@ const (
 	Attr_NetworkName                                 = "network_name"
 	Attr_NetworkPorts                                = "network_ports"
 	Attr_Networks                                    = "networks"
+	Attr_NetworkSecurityGroupID                      = "network_security_group_id"
+	Attr_NetworkSecurityGroupMembers                 = "network_security_group_members"
+	Attr_NetworkSecurityGroupRules                   = "network_security_group_rules"
```
Comment (michaelkad) [inline ibm/service/power/ibm_pi_constants.go:287]:
```suggestion
	Attr_Members                 = "members"
	Attr_Rules                   = "rules"
```
Those should be

File: ibm/service/power/ibm_pi_constants.go [hunk 20]
```diff
@@ -447,6 +481,12 @@ const (
 	vSCSI                     = "vSCSI"
 	Warning                   = "WARNING"
 
+	// Protocol Type
+	All  = "all"
+	ICMP = "icmp"
+	TCP  = "tcp"
+	UDP  = "udp"
```
Comment (michaelkad) [inline ibm/service/power/ibm_pi_constants.go:487]: Those should go under the // Allowed values

File: ibm/service/power/resource_ibm_pi_instance.go [hunk 21]
```diff
@@ -1630,6 +1631,7 @@ func expandDeploymentTarget(dt []interface{}) *models.DeploymentTarget {
 	}
 	return dtexpanded
 }
+
```
Comment (michaelkad) [inline ibm/service/power/resource_ibm_pi_instance.go:1634]: Those changes can go on a different PR

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go [hunk 22]
```diff
@@ -0,0 +1,654 @@
...
+		},
+		Schema: map[string]*schema.Schema{
+			// Arguments
+			Arg_Action: {
+				ConflictsWith: []string{Arg_NetworkSecurityGroupRuleID},
+				Description:   "The action to take if the rule matches network traffic.",
+				ForceNew:      true,
+				Optional:      true,
+				Type:          schema.TypeString,
+				ValidateFunc:  validate.ValidateAllowedStringValues([]string{"allow", "deny"}),
```
Comment (michaelkad) [inline ibm/service/power/resource_ibm_pi_network_security_group_rule.go:40]: Could you add those to the const file

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go [hunk 23]
```diff
@@ -0,0 +1,654 @@
...
+						Attr_ID: {
+							Description: "The ID of the remote network address group or network security group the rules apply to. Not required for default-network-address-group.",
+							Optional:    true,
+							Type:        schema.TypeString,
+						},
+						Attr_Type: {
+							Description:  "The type of remote group (MAC addresses, IP addresses, CIDRs, external CIDRs) that are the originators of rule's network traffic to match.",
+							Optional:     true,
+							Type:         schema.TypeString,
+							ValidateFunc: validate.ValidateAllowedStringValues([]string{"default-network-address-group", "network-address-group", "network-security-group"}),
```
Comment (michaelkad) [inline ibm/service/power/resource_ibm_pi_network_security_group_rule.go:138]: Could you add them to the const file?
`DefaultNAG, NAG, NSG`

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go [hunk 24]
```diff
@@ -0,0 +1,654 @@
...
+					},
+				},
+				Type: schema.TypeList,
+			},
+			Attr_UserTags: {
+				Computed:    true,
+				Description: "The user tags associated with this resource.",
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Type:        schema.TypeList,
+			},
```
Comment (michaelkad) [inline ibm/service/power/resource_ibm_pi_network_security_group_rule.go:332]: This can be removed for now.

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go [hunk 25]
```diff
@@ -0,0 +1,654 @@
...
+
+func ResourceIBMPINetworkSecurityGroupRule() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceIBMPINetworkSecurityGroupRuleCreate,
+		ReadContext:   resourceIBMPINetworkSecurityGroupRuleRead,
+		DeleteContext: resourceIBMPINetworkSecurityGroupRuleDelete,
+		Importer:      &schema.ResourceImporter{},
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(60 * time.Minute),
+			Delete: schema.DefaultTimeout(60 * time.Minute),
```
Comment (michaelkad) [inline ibm/service/power/resource_ibm_pi_network_security_group_rule.go:30]: I think 10 min is reasonable here for the default

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go [hunk 26]
```diff
@@ -0,0 +1,654 @@
...
+		rules := []map[string]interface{}{}
+		for _, rule := range networkSecurityGroup.Rules {
+			ruleMap := networkSecurityGroupRuleToMap(rule)
+			rules = append(rules, ruleMap)
+		}
+		d.Set(Attr_Rules, rules)
+	}
+	if len(Attr_UserTags) > 0 {
+		d.Set(Attr_UserTags, networkSecurityGroup.UserTags)
+	}
```
Comment (michaelkad) [inline ibm/service/power/resource_ibm_pi_network_security_group_rule.go:434]: Can be removed

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go [hunk 27]
```diff
@@ -0,0 +1,654 @@
...
+
+	d.Set(Attr_NetworkSecurityGroupID, networkSecurityGroup.ID)
+	if len(networkSecurityGroup.Rules) > 0 {
+		rules := []map[string]interface{}{}
+		for _, rule := range networkSecurityGroup.Rules {
+			ruleMap := networkSecurityGroupRuleToMap(rule)
+			rules = append(rules, ruleMap)
+		}
+		d.Set(Attr_Rules, rules)
+	}
```
Comment (michaelkad) [inline ibm/service/power/resource_ibm_pi_network_security_group_rule.go:431]:
```suggestion
	if len(networkSecurityGroup.Rules) > 0 {
		rules := []map[string]interface{}{}
		for _, rule := range networkSecurityGroup.Rules {
			ruleMap := networkSecurityGroupRuleToMap(rule)
			rules = append(rules, ruleMap)
		}
		d.Set(Attr_Rules, rules)
	}else{
	     d.Set(Attr_Rules, nil)
	}
```

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go [hunk 28]
```diff
@@ -0,0 +1,654 @@
...
+
+func isWaitForIBMPINetworkSecurityGroupRuleAdd(ctx context.Context, client *instance.IBMPINetworkSecurityGroupClient, id, ruleID string, timeout time.Duration) (interface{}, error) {
+
+	stateConf := &retry.StateChangeConf{
+		Pending:    []string{State_Pending},
+		Target:     []string{State_Available},
+		Refresh:    isIBMPINetworkSecurityGroupRuleAddRefreshFunc(client, id, ruleID),
+		Timeout:    timeout,
+		Delay:      10 * time.Second,
+		MinTimeout: 10 * time.Minute,
```
Comment (michaelkad) [inline ibm/service/power/resource_ibm_pi_network_security_group_rule.go:477]:
```suggestion
		MinTimeout: 10 * time.Minute,
```
This can be reduce to 60 sec

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go [hunk 29]
```diff
@@ -0,0 +1,654 @@
...
+
+func isWaitForIBMPINetworkSecurityGroupRuleRemove(ctx context.Context, client *instance.IBMPINetworkSecurityGroupClient, id, ruleID string, timeout time.Duration) (interface{}, error) {
+
+	stateConf := &retry.StateChangeConf{
+		Pending:    []string{State_Pending},
+		Target:     []string{State_Removed},
+		Refresh:    isIBMPINetworkSecurityGroupRuleRemoveRefreshFunc(client, id, ruleID),
+		Timeout:    timeout,
+		Delay:      10 * time.Second,
+		MinTimeout: 10 * time.Minute,
```
Comment (michaelkad) [inline ibm/service/power/resource_ibm_pi_network_security_group_rule.go:511]:
```suggestion
		MinTimeout: 10 * time.Minute,
```
This can be reduce to 60 sec

File: ibm/service/power/ibm_pi_constants.go (hunk 14 above)
Comment (ismirlia) [reply ibm/service/power/ibm_pi_constants.go:101]: updated

File: ibm/service/power/resource_ibm_pi_instance.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (ismirlia) [reply ibm/service/power/resource_ibm_pi_instance.go:1634]: sorry I think that was unintended

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go (hunk 25 above)
Comment (ismirlia) [reply ibm/service/power/resource_ibm_pi_network_security_group_rule.go:30]: updated

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go (hunk 26 above)
Comment (ismirlia) [reply ibm/service/power/resource_ibm_pi_network_security_group_rule.go:434]: updated

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go (hunk 27 above)
Comment (ismirlia) [reply ibm/service/power/resource_ibm_pi_network_security_group_rule.go:431]: updated

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go (hunk 29 above)
Comment (ismirlia) [reply ibm/service/power/resource_ibm_pi_network_security_group_rule.go:511]: updated

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go (hunk 28 above)
Comment (ismirlia) [reply ibm/service/power/resource_ibm_pi_network_security_group_rule.go:477]: updated

File: ibm/service/power/ibm_pi_constants.go [hunk 30]
```diff
@@ -20,10 +20,12 @@ const (
 	Arg_DeploymentTarget                    = "pi_deployment_target"
 	Arg_DeploymentType                      = "pi_deployment_type"
 	Arg_Description                         = "pi_description"
+	Arg_DestinationPorts                    = "pi_destination_ports"
 	Arg_DhcpID                              = "pi_dhcp_id"
 	Arg_DhcpName                            = "pi_dhcp_name"
 	Arg_DhcpSnatEnabled                     = "pi_dhcp_snat_enabled"
 	Arg_DnsServer                           = "pi_dns_server"
+	Arg_Flag                                = "pi_flag"
```
[the remark this answers is not in the document]
Comment (ismirlia) [reply ibm/service/power/ibm_pi_constants.go:28]: updated

File: ibm/service/power/ibm_pi_constants.go [hunk 31]
```diff
@@ -32,6 +34,8 @@ const (
 	Arg_IBMiCSS                             = "pi_ibmi_css"
 	Arg_IBMiPHA                             = "pi_ibmi_pha"
 	Arg_IBMiRDSUsers                        = "pi_ibmi_rds_users"
+	Arg_ICMPTypes                           = "pi_icmp_types"
```
[the remark this answers is not in the document]
Comment (ismirlia) [reply ibm/service/power/ibm_pi_constants.go:37]: updated

File: ibm/service/power/ibm_pi_constants.go [hunk 32]
```diff
@@ -32,6 +34,8 @@ const (
...
+	Arg_ID                                  = "pi_id"
```
[the remark this answers is not in the document]
Comment (ismirlia) [reply ibm/service/power/ibm_pi_constants.go:38]: updated

File: ibm/service/power/ibm_pi_constants.go (hunk 20 above)
Comment (ismirlia) [reply ibm/service/power/ibm_pi_constants.go:484]: moved

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go (hunk 24 above)
Comment (ismirlia) [reply ibm/service/power/resource_ibm_pi_network_security_group_rule.go]: removed

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go (hunk 23 above)
Comment (ismirlia) [reply ibm/service/power/resource_ibm_pi_network_security_group_rule.go]: updated

File: ibm/service/power/resource_ibm_pi_network_security_group_rule.go (hunk 22 above)
Comment (ismirlia) [reply ibm/service/power/resource_ibm_pi_network_security_group_rule.go]: updated

**Review by Alexander-Kita — APPROVED** [review]

**Review by michaelkad — APPROVED** [review]

File: website/docs/r/pi_network_security_group_rule.html.markdown [hunk 33]
```diff
@@ -0,0 +1,137 @@
...
+      pi_destination_ports {
+        minimum = 1200
+        maximum = 37466
+      }
+      pi_source_ports {
+        minimum = 1000
+        maximum = 19500
+      }
+      pi_protocol {
+        icmp_types = []
```
Comment (Alexander-Kita) [inline website/docs/r/pi_network_security_group_rule.html.markdown:29]: This argument needs to be updated everywhere it is present in the documentation, including no longer having to provide an empty array.

Comment (ismirlia) [reply website/docs/r/pi_network_security_group_rule.html.markdown]: Updated

**Comment by ismirlia** [comment]
Closing as https://github.com/powervs-ibm/terraform-provider-ibm/pull/157 has incorporated my changes.
